### PR TITLE
noise: get rid of the intricate locking/atomics

### PR DESCRIPTION
**Pull request by Noah-Kennedy**
The existing scheme does not work, and getting rid of it allows us to put a lock at top-level, eliminating a bunch of races.

### CODE REVIEW

**Comment by jeff-hiner** [comment]
This definitely looks reasonable.

File: boringtun/src/noise/timers.rs [hunk 1]
```diff
@@ -322,14 +305,12 @@ impl Tunn {
...
     }
 
-    pub fn time_since_last_handshake(&self) -> Option<Duration> {
-        let current_session = self.current.load(Ordering::Acquire);
-        if self.sessions[current_session % super::N_SESSIONS]
-            .read()
-            .is_some()
-        {
+    pub(super) fn time_since_last_handshake(&self) -> Option<Duration> {
+        tracing::error!(?self.sessions, "Sessions");
```
Comment (schultetwin1) [inline boringtun/src/noise/timers.rs]: Should this still be here as `tracing::error` or was it just added for debugging?

File: boringtun/src/noise/mod.rs [hunk 2]
```diff
@@ -336,16 +448,13 @@ impl Tunn {
             remote_idx = p.sender_idx
         );
 
-        let session = {
-            let mut handshake = self.handshake.lock();
-            handshake.receive_handshake_response(p)?
-        };
+        let session = { self.handshake.receive_handshake_response(p)? };
```
Comment (schultetwin1) [inline boringtun/src/noise/mod.rs:451]: Do we still need the `{}` brackets surrounding the expression here?

**Review by schultetwin1 — COMMENTED** [review]
One overarching question.

It looks like this change will serialize the processing of packets, even packets tied to different WireGuard sessions. This used to not be the case. Are you concerned about any perf issues that may arrive from this change?

File: boringtun/src/noise/timers.rs (hunk 1 above)
Comment (schultetwin1) [reply boringtun/src/noise/timers.rs]: Looks like it was removed during my review!

**Comment by Noah-Kennedy** [comment]
> One overarching question.
> 
> It looks like this change will serialize the processing of packets, even packets tied to different WireGuard sessions. This used to not be the case. Are you concerned about any perf issues that may arrive from this change?

So, this is slightly incorrect. "Session" is an ambiguous term here, because in wireguard, it actually refers to handshakes and which keys are currently valid for encrypting data.

This change causes code to be serialized for operations involving a single peer. In the future, we can add more fine-grained locking and synchronization to improve this. That said, I am not concerned about this for real-world use case, as it is not often that you actually have simultaneous reading and writing of IP packets to a specific peer.

File: boringtun/src/noise/mod.rs (hunk 2 above)
Comment (Noah-Kennedy) [reply boringtun/src/noise/mod.rs:451]: No, this is a mistake, I meant to remove this.

File: boringtun/src/noise/timers.rs (hunk 1 above)
Comment (Noah-Kennedy) [reply boringtun/src/noise/timers.rs]: Yeah, it was added for debugging the underflow issue.

File: boringtun/src/noise/mod.rs (hunk 2 above)
Comment (Noah-Kennedy) [reply boringtun/src/noise/mod.rs]: Done.

**Comment by schultetwin1** [comment]
> So, this is slightly incorrect. "Session" is an ambiguous term here, because in wireguard, it actually refers to handshakes and which keys are currently valid for encrypting data.

Yep, understood that the WireGuard spec has a specific definition for session. That definition of session is what I was referring to by "WireGuard Session".

Asked another way, we were previously able to both encapsulate and decapsulate multiple packets at the same time for one peer. Even if those packets had different ephemeral keys associated with them. With this change we have serialized the process so we can only do one at a time.

Answering my own question: On the client side, I think that this serialization is ok because we only process packets one at a time anyways. Since the reading in of packets is serialized and the writing out of packets is serialized, putting a lock in here is ok. Though, as you pointed out, we can no longer send and receive at the same time.

File: boringtun/src/noise/mod.rs [hunk 3]
```diff
@@ -307,14 +422,11 @@ impl Tunn {
             remote_idx = p.sender_idx
         );
 
-        let (packet, session) = {
-            let mut handshake = self.handshake.lock();
-            handshake.receive_handshake_initialization(p, dst)?
-        };
+        let (packet, session) = { self.handshake.receive_handshake_initialization(p, dst)? };
```
Comment (jeff-hiner) [inline boringtun/src/noise/mod.rs:425]: This no longer needs brackets.

File: boringtun/src/noise/timers.rs [hunk 4]
```diff
@@ -41,197 +42,179 @@ pub enum TimerName {
...
 #[derive(Debug)]
 pub struct Timers {
     /// Is the owner of the timer the initiator or the responder for the last handshake?
-    is_initiator: AtomicBool,
+    is_initiator: bool,
     /// Start time of the tunnel
     time_started: Instant,
-    timers: [Timer; TimerName::Top as usize],
-    pub(super) session_timers: [Timer; super::N_SESSIONS],
+    timers: [Duration; TimerName::Top as usize],
```
Comment (jeff-hiner) [inline boringtun/src/noise/timers.rs:51]: Yes! Thank you!

**Review by jeff-hiner — APPROVED** [review]
Excellent work. Thank you!

**Review by schultetwin1 — APPROVED** [review]
Nice work!

**Pull request merged**